File: backend/app/schemas/ab_tests.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field, field_validator


class ABTestPhotoIn(BaseModel):
    order: int
    file_url: str


class ABTestCreateRequest(BaseModel):
    account_id: int
    nm_id: int
    product_card_id: int | None = None
    card_id: int | None = None
    title: str
    from_main: bool = False
    main_photo_url: str | None = None
    max_slots: int = 5
    keep_winner_as_main: bool = True
    delete_test_photos: bool = True
    photos: list[ABTestPhotoIn] = Field(default_factory=list)
    photos_count: int = 0
    preview_confirmed: bool = False
    confirm: bool = False

    @field_validator("photos")
    @classmethod
    def _validate_photos(cls, value: list[ABTestPhotoIn]) -> list[ABTestPhotoIn]:
        if len(value) < 2:
            raise ValueError("At least two photo variants are required")
        return value


class ABTestUpdateRequest(BaseModel):
    account_id: int
    id_company: int | None = None
    company_id: int | None = None
    nm_id: int
    product_card_id: int | None = None
    card_id: int | None = None
    title: str
    title_changed: bool = False
    from_main: bool = False
    max_slots: int = 5
    keep_winner_as_main: bool = True
    delete_test_photos: bool = True
    photos_count: int = 0
    views_per_photo: int
    cpm: int
    spend_rub: int = 0
    estimated_spend_rub: int | None = None
    auto_deposit: bool = True
    deposit_rub: int | None = None
    payment_source: str | None = None
    use_promo_bonus: bool | None = None
    photos: list[ABTestPhotoIn] = Field(default_factory=list)
    preview_confirmed: bool = False
    confirm: bool = False

    @field_validator("views_per_photo")
    @classmethod
    def _validate_views(cls, value: int) -> int:
        if int(value) < 1000:
            raise ValueError("views_per_photo must be at least 1000")
        return value

    @field_validator("cpm")
    @classmethod
    def _validate_cpm(cls, value: int) -> int:
        if int(value) <= 0:
            raise ValueError("cpm must be positive")
        return value


class ABTestPhotoOut(BaseModel):
    order: int
    file_url: str
    wb_url: str | None = None
    preview_url: str | None = None
    shows: int = 0
    clicks: int = 0
    ctr: float = 0.0
    is_winner: bool = False
    winner_score: float | None = None
    winner_score_confidence: float | None = None
    winner_score_conversion_source: str | None = None
    winner_score_reason: str | None = None


class ABTestCompanyOut(BaseModel):
    id_company: int
    company_id: int | None = None
    wb_advert_id: int | None = None
    account_id: int
    nm_id: int
    product_card_id: int | None = None
    card_id: int | None = None
    title: str
    status: str
    spend_rub: int
    estimated_spend_rub: int = 0
    winner_decision: str | None = None
    views_per_photo: int
    photos_count: int
    current_photo_order: int
    winner_photo_order: int | None = None
    last_error: str | None = None
    can_start: bool = False
    can_stop: bool = False
    started_at: datetime | None = None
    finished_at: datetime | None = None
    photos: list[ABTestPhotoOut] = Field(default_factory=list)


class ABTestPageOut(BaseModel):
    items: list[ABTestCompanyOut]
    pagination: dict[str, int]


class ABTestBalanceOut(BaseModel):
    balance: int = 0
    promo_bonus_rub: int = 0
    raw: Any | None = None

File: frontend/src/lib/ab-tests.ts
import { api, apiList } from "@/lib/api";

export type ABTestStatus = "running" | "pending" | "finished" | "failed";

export interface ABTestPhoto {
  order: number;
  file_url: string;
  wb_url?: string | null;
  preview_url?: string | null;
  shows: number;
  clicks: number;
  ctr: number;
  is_winner: boolean;
  winner_score?: number | null;
  winner_score_confidence?: number | null;
  winner_score_conversion_source?: string | null;
  winner_score_reason?: string | null;
}

export interface ABTestCompany {
  id_company: number;
  company_id?: number | null;
  wb_advert_id?: number | null;
  account_id: number;
  nm_id: number;
  product_card_id?: number | null;
  card_id?: number | null;
  title: string;
  status: string;
  spend_rub: number;
  estimated_spend_rub: number;
  winner_decision?: string | null;
  views_per_photo: number;
  photos_count: number;
  current_photo_order: number;
  winner_photo_order?: number | null;
  last_error?: string | null;
  can_start?: boolean;
  can_stop?: boolean;
  started_at?: string | null;
  finished_at?: string | null;
  photos: ABTestPhoto[];
}

export interface ProductOption {
  id: number;
  nm_id: number;
  vendor_code?: string | null;
  title?: string | null;
  brand?: string | null;
  photos?: unknown;
}

export function fetchABTests(
  accountId: number,
  status: ABTestStatus,
  params?: { limit?: number; offset?: number },
) {
  return api<{ items: ABTestCompany[]; pagination: Record<string, number> }>(
    `/promotion/${status}`,
    {
      query: {
        account_id: accountId,
        page: Math.floor((params?.offset ?? 0) / (params?.limit ?? 50)) + 1,
        page_size: params?.limit ?? 50,
      },
    },
  );
}

export function fetchABTestBalance(accountId: number) {
  return api<{ balance: number; promo_bonus_rub: number; raw?: unknown }>(
    "/promotion/balance",
    {
      query: { account_id: accountId },
    },
  );
}

export function createABTestCompany(
  accountId: number,
  payload: Record<string, unknown>,
) {
  return api<Record<string, unknown>>("/promotion/create_company", {
    method: "POST",
    body: { ...payload, account_id: accountId },
  });
}

export function updateABTestCompany(
  accountId: number,
  payload: Record<string, unknown>,
) {
  return api<Record<string, unknown>>("/promotion/update", {
    method: "POST",
    body: { ...payload, account_id: accountId },
  });
}

export function startABTestCompany(
  accountId: number,
  companyId: number,
  options?: { confirm?: boolean },
) {
  return api<Record<string, unknown>>(`/promotion/company/${companyId}/start`, {
    method: "POST",
    query: { account_id: accountId, confirm: options?.confirm || undefined },
  });
}

export function stopABTestCompany(
  accountId: number,
  companyId: number,
  options?: { confirm?: boolean },
) {
  return api<Record<string, unknown>>(`/promotion/company/${companyId}/stop`, {
    method: "POST",
    query: { account_id: accountId, confirm: options?.confirm || undefined },
  });
}

export function fetchABTestStats(accountId: number, companyId: number) {
  return api<ABTestCompany>(`/promotion/company/${companyId}/stats`, {
    query: { account_id: accountId },
  });
}

export function fetchProductsForABTest(accountId: number, search: string) {
  return apiList<ProductOption>("/products", {
    query: {
      account_id: accountId,
      search: search || undefined,
      limit: 80,
      sort_by: "updated_at_wb",
      sort_dir: "desc",
    },
  });
}

export function uploadABTestPhoto(file: File, accountId: number) {
  const formData = new FormData();
  formData.append("file", file);
  formData.append("asset_type", "custom");
  formData.append("name", file.name || "A/B variant");
  return api<{
    id: number;
    file_url?: string;
    url?: string;
    image_url?: string;
    file_name?: string;
  }>("/photo/assets/upload", {
    method: "POST",
    query: { account_id: accountId },
    formData,
  });
}

export function extractProductPhotos(product: ProductOption): string[] {
  const raw = product.photos;
  const out: string[] = [];
  const push = (value: unknown) => {
    if (typeof value === "string" && value.trim()) out.push(value.trim());
    else if (value && typeof value === "object") {
      const obj = value as Record<string, unknown>;
      for (const key of [
        "big",
        "url",
        "full",
        "c516x688",
        "c246x328",
        "square",
      ]) {
        if (typeof obj[key] === "string" && String(obj[key]).trim()) {
          out.push(String(obj[key]).trim());
          break;
        }
      }
    }
  };
  if (Array.isArray(raw)) raw.forEach(push);
  else push(raw);
  return Array.from(new Set(out));
}
